metadata-helper/src/signatures.rs

```diff
@@ -0,0 +1,119 @@
+use crate::AppState;
+
+use actix_files::NamedFile;
+use actix_web::http::header;
+use actix_web::HttpRequest;
+use commons::tracing::get_tracer;
+use commons::{self, Fallible, GraphError};
+use opentelemetry::trace::{mark_span_as_active, Tracer};
+use prometheus::{histogram_opts, Histogram, IntCounterVec, Opts, Registry};
+use std::path::PathBuf;
+
+lazy_static! {
+    static ref SIGNATURES_INCOMING_REQS: IntCounterVec = IntCounterVec::new(
+        Opts::new("signatures_incoming_requests_total",
+        "Total number of incoming HTTP client request"),
+        &["uri_path"]
+    )
+    .unwrap();
+    // Histogram with custom bucket values for serving latency metric (in seconds), values are picked based on monthly data
+    static ref SIGNATURES_SERVE_HIST: Histogram = Histogram::with_opts(histogram_opts!(
+        "signatures_serve_duration_seconds",
+        "HTTP graph serving latency in seconds",
+        vec![0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25, 0.5, 0.75, 1.0, 5.0]
+    ))
+    .unwrap();
+}
+
+/// Register relevant metrics to a prometheus registry.
+pub(crate) fn register_metrics(registry: &Registry) -> Fallible<()> {
+    commons::register_metrics(registry)?;
+    registry.register(Box::new(SIGNATURES_INCOMING_REQS.clone()))?;
+    registry.register(Box::new(SIGNATURES_SERVE_HIST.clone()))?;
+    Ok(())
+}
+
+/// Serve Cincinnati graph requests.
+pub(crate) async fn index(
+    req: HttpRequest,
+    app_data: actix_web::web::Data<AppState>,
+    digest: String,
+    signature: String,
+) -> Result<NamedFile, GraphError> {
+    debug!("{}, {}", digest, signature);
+    _index(&req, app_data)
+        .await
+        .map_err(|e| api_response_error(&req, e))
+}
+
+async fn _index(
+    req: &HttpRequest,
+    app_data: actix_web::web::Data<AppState>,
+) -> Result<NamedFile, GraphError> {
+    let span = get_tracer().start("index");
+    let _active_span = mark_span_as_active(span);
+
+    let path = req.uri().path();
+    SIGNATURES_INCOMING_REQS.with_label_values(&[path]).inc();
+
+    let timer = SIGNATURES_SERVE_HIST.start_timer();
+
+    let params = req.match_info();
+    let algo = params.get("ALGO").unwrap();
+    let digest = params.get("DIGEST").unwrap();
+    let signature = params.get("SIGNATURE").unwrap();
+
+    let signatures_data_path = app_data.signatures_dir.clone();
+    let mut signature_path = PathBuf::from(signatures_data_path);
+    signature_path.push(&format!("{}/{}/{}", algo, digest, signature));
```

Comment:
Looking at RFC 3986's sections [3.3][1] and [5.2.4][2], I don't think we need to worry about `..` path segments being used to access locations outside of `signatures_data_path`, but it's probably worth testing just to make sure.

[1]: https://www.rfc-editor.org/rfc/rfc3986#section-3.3
[2]: https://www.rfc-editor.org/rfc/rfc3986#section-5.2.4